File: 0713_subarray_product_less_than_k/num_subarray.py
from typing import List


def numSubarrayProductLessThanK(nums: List[int], k: int) -> int:
    """
    Time: O(n)
    Space: O(1)
    """
    if k <= 1:
        return 0
    res = left = 0
    p = 1
    for right, e in enumerate(nums):
        p *= e
        while p >= k:
            p /= nums[left]
            left += 1
        res += right - left + 1
    return res

File: 0713_subarray_product_less_than_k/num_subarray.java
class Solution {
    // Time: O(n)
    // Space: O(1)
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        if (k <= 1) {
            return 0;
        }
        int res = 0, l = 0, p = 1;
        for (int i = 0; i < nums.length; i++) {
            p *= nums[i];
            while (p >= k) {
                p /= nums[l++];
            }
            res += i - l + 1;
        }
        return res;
    }
}